contracts/defi/provider/MakerProvider.sol

```diff
@@ -0,0 +1,817 @@
+        invokeWallet(_wallet, daiToken, 0, abi.encodeWithSelector(ERC20_APPROVE, address(_makerCdp), _amount));
+        // Approve MKR to let wipe() pay the MKR governance fee
+        invokeWallet(_wallet, mkrToken, 0, abi.encodeWithSelector(ERC20_APPROVE, address(_makerCdp), mkrFee));
+        // repay DAI debt and MKR governance fee
+        invokeWallet(_wallet, address(_makerCdp), 0, abi.encodeWithSelector(CDP_WIPE, _cup, _amount));
+        // emit CdpUpdated
+        emit CdpUpdated(address(_wallet), _cup, pethCollateral(_cup, _makerCdp), daiDebt(_cup, _makerCdp));    
+    }
+
+    /**
+     * @dev Lets the owner of a CDP close their CDP. The method will 1) repay all debt 
+     * and governance fee, 2) free all collateral, and 3) delete the CDP.
+     * @param _wallet The target wallet
+     * @param _cup The id of the CDP.
+     * @param _makerCdp The Maker CDP contract
+     * @param _uniswapFactory The Uniswap Factory contract.
+     */
+    function closeCdp(
+        BaseWallet _wallet, 
+        bytes32 _cup,
+        IMakerCdp _makerCdp,
+        UniswapFactory _uniswapFactory
+    ) 
+        internal
+    {
+        // repay all debt (in DAI) + stability fee (in DAI) + governance fee (in MKR)
+        removeDebt(_wallet, _cup, daiDebt(_cup, _makerCdp), _makerCdp, _uniswapFactory);
+        // free all ETH collateral
+        removeCollateral(_wallet, _cup, pethCollateral(_cup, _makerCdp), _makerCdp);
+        // shut the CDP
+        invokeWallet(_wallet, address(_makerCdp), 0, abi.encodeWithSelector(CDP_SHUT, _cup));
+        // emit CdpClosed
+        emit CdpClosed(address(_wallet), _cup);    
+    }
+
+    /* Convenience methods */
+
+    /**
+     * @dev Returns the amount of PETH collateral locked in a CDP.
+     * @param _cup The id of the CDP.
+     * @param _makerCdp The Maker CDP contract
+     * @return the amount of PETH locked in the CDP.
+     */
+    function pethCollateral(bytes32 _cup, IMakerCdp _makerCdp) public view returns (uint256) { 
+        return _makerCdp.ink(_cup);
+    }
+
+    /**
+     * @dev Returns the amount of DAI debt (including the stability fee if non-zero) drawn from a CDP.
+     * @param _cup The id of the CDP.
+     * @param _makerCdp The Maker CDP contract
+     * @return the amount of DAI drawn from the CDP.
+     */
+    function daiDebt(bytes32 _cup, IMakerCdp _makerCdp) public returns (uint256) { 
+        return _makerCdp.tab(_cup);
+    }
+
+    /**
+     * @dev Indicates whether a CDP is above the liquidation ratio.
+     * @param _cup The id of the CDP.
+     * @param _makerCdp The Maker CDP contract
+     * @return false if the CDP is in danger of being liquidated.
+     */
+    function isSafe(bytes32 _cup, IMakerCdp _makerCdp) public returns (bool) { 
+        return _makerCdp.safe(_cup);
+    }
+
+    /**
+     * @dev Checks if a CDP exists.
+     * @param _cup The id of the CDP.
+     * @return false if the CDP is in danger of being liquidated.
```

Comment:
`ink` is just the amount of collateral in the CDP. Why does `ink == 0` implies a danger of liquidation?

Reply:
if `tab==0` (no debt) and `ink==0` (no collateral), then the CDP is safe

Reply:
There is an error in the comment (lousy cut and paste...). Here we only check if the CDP exists, and we do so by checking if there is collateral (no collateral means no CDP). 

I've updated the comment according to the method :-)